ouroboros-network-framework/src/Ouroboros/Network/Subscription/Dns.hs

```diff
@@ -115,143 +116,104 @@ dnsResolve tracer getSeed withResolverFn peerStatesVar beforeConnect (DnsSubscri
-        addrs <- atomically $ takeTMVar addrsVar
-        listTargets (Left addrs) (Left [])
-
-    -- Peek at the result for one family.
-    listTargets (Right addrsVar) (Left a) = do
-        addrs_m <- atomically $ tryTakeTMVar addrsVar
-        case addrs_m of
-             Just addrs -> listTargets (Left addrs) (Left a)
-             Nothing    -> listTargets (Left a) (Right addrsVar)
+    -- Creates a subscription target from an optional first socket and a tail
+    targetCons
+      :: Socket.SockAddr
+      -> m (Maybe (Socket.SockAddr, SubscriptionTarget m Socket.SockAddr))
+      -> m (Maybe (Socket.SockAddr, SubscriptionTarget m Socket.SockAddr))
+    targetCons addr next = do
+      b <- runBeforeConnect peerStatesVar beforeConnect addr
+      if b
+        then return $ Just (addr, SubscriptionTarget next)
+        else next
+
+    -- Takes the result of a thread, returning an optional first socket in the subscription target result,
+    -- then calls the given function to get the tail
+    handleThreadResult
+      :: Either SomeException [Socket.SockAddr]
+      -> ([Socket.SockAddr] -> m (Maybe (Socket.SockAddr, SubscriptionTarget m Socket.SockAddr)))
+      -> m (Maybe (Socket.SockAddr, SubscriptionTarget m Socket.SockAddr))
+    handleThreadResult (Left e) cont = do
+      traceWith tracer $ DnsTraceLookupException e
+      cont []
+    handleThreadResult (Right []) cont = cont []
+    handleThreadResult (Right (addr:addrs)) cont = targetCons addr $ cont addrs
+
+    -- Called when a thread is still running, and the other finished already
+    -- Cycles between trying to get a result from the running thread, and the results of the finished thread
+    -- If results of the finished thread are exhausted, wait until the running thread completes
+    threadTargetCycle
+      :: Async m [Socket.SockAddr]
+      -> [Socket.SockAddr]
+      -> m (Maybe (Socket.SockAddr, SubscriptionTarget m Socket.SockAddr))
+    threadTargetCycle asyn [] = do
+      result <- waitCatch asyn
+      handleThreadResult result $ targetCycle []
+    threadTargetCycle asyn a@(addr : addrs) = do
+      result <- poll asyn
+      case result of
+        -- The running thread finished, handle the result, then cycle over all results
+        Just r -> handleThreadResult r $ targetCycle a
+        -- The running thread is still going, emit an address of the finished thread, then check again
+        Nothing -> targetCons addr $ threadTargetCycle asyn addrs
+
+    -- Called when both threads exited and we know the results of both.
+    -- Returns a subscription target that cycles between the results until both results are exhausted
+    targetCycle
+      :: [Socket.SockAddr]
+      -> [Socket.SockAddr]
+      -> m (Maybe (Socket.SockAddr, SubscriptionTarget m Socket.SockAddr))
+    targetCycle [] [] = return Nothing
+    targetCycle [] ipvB = targetCycle ipvB []
+    targetCycle (addr : addrs) ipvB = targetCons addr $ targetCycle ipvB addrs
```

Comment:
This would change the behavior to not cycle between the address families anymore. Currently it goes ipv4, ipv6, ipv4, ipv6, ... (or the other way around if ipv6 was first)

Reply:
Right, the recursive call to `targetCycle` switches the order of arguments which is quite easy to miss.  What about this then:
```
targetCycle :: [Socket.SockAdr] -> [Socket.SockAddr] -> m (Maybe ...)
targetCycle as bs = go (as `interleave` bs)
  where
    go []       = return Nothing
    go (a : as) = targetCons a (go as)
    
    interleave :: [a] -> [a] -> [a]
    interleave []       bs = bs
    interleave (a : as) bs = a : interleave bs as
```

This will make it more explicit and harder to miss.

Reply:
One thing that I still don't like is how `targetCons` is used.  We should use `First` & `foldMap` instead of `targetCons`:
```
targetCycle :: (a -> Maybe b) -> [a] -> [a] -> Maybe b
targetCycle f as bs = getFirst . foldMap (First . f) (as `interleave` bs)
  where
    interleave = ... -- as above
```
and we'd pass `runBeforeConnect` as `f`.

Reply:
I think the mental overhead for this is a bit too high, I'd prefer the explicit `go` function instead.

Reply:
Went for that now